File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/services/S3FileService.java
package uk.gov.justice.digital.hmpps.riskprofiler.services;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.ListObjectsV2Result;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import com.amazonaws.util.IOUtils;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.time.ZoneId;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Component
@Slf4j
@ConditionalOnProperty(name = "file.process.type", havingValue = "s3")
public class S3FileService implements FileService {

    private Map<String, AmazonS3> bucketAccountMap;

    public S3FileService(@Qualifier("s3Client") AmazonS3 s3Client, @Qualifier("viperS3Client") AmazonS3 viperS3client,
                         @Value("${bucket.account.map}") List<String> clientList) {

        bucketAccountMap = getBucketClientMap(clientList).entrySet().stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> e.getValue().equals("s3Client") ? s3Client : viperS3client));

    }

    private Map<String, String> getBucketClientMap(@Value("${bucket.account.map}") List<String> clientList) {
        return clientList.stream()
                .collect(Collectors.toMap(
                        v -> StringUtils.split(v, "|")[0],
                        v -> StringUtils.split(v, "|")[1],
                        (v1, v2) -> {
                            log.warn("duplicate key found {}", v1);
                            return v1;
                        }));
    }


    public PendingFile getLatestFile(String fileLocation) {

        var s3Result = getObjectSummaries(fileLocation);

        log.info("Found {} objects in {}", s3Result.getObjects().size(), fileLocation);
        return s3Result.getObjects().stream()
                .max(Comparator.comparing(S3ObjectSummary::getLastModified))
                .map(o -> {
                    try {
                        var s3Object = s3Result.getAmazonS3Client().getObject(s3Result.getBucketName(), o.getKey());
                        return PendingFile.builder()
                                .fileName(o.getKey())
                                .fileTimestamp(o.getLastModified().toInstant()
                                        .atZone(ZoneId.systemDefault())
                                        .toLocalDateTime())
                                .data(IOUtils.toByteArray(s3Object.getObjectContent()))
                                .build();

                    } catch (IOException e) {
                        return null;
                    }
                }).orElse(null);
    }

    @Override
    public void deleteHistoricalFiles(String fileLocation) {
       var s3ObjectResult = getObjectSummaries(fileLocation);

        log.info("Found {} data files for data housekeeping in {}", s3ObjectResult.getObjects().size(), fileLocation);

        s3ObjectResult.getObjects().stream().sorted(Comparator.comparing(S3ObjectSummary::getLastModified).reversed()).skip(2).forEach(o -> {
            s3ObjectResult.getAmazonS3Client().deleteObject(s3ObjectResult.getBucketName(), o.getKey());
            log.info("Deleted s3 data file: {} from bucket {}", o.getKey(), s3ObjectResult.getBucketName());
        });
    }

    private ObjectSummaryResult getObjectSummaries(String fileLocation){
        var bucketAndPrefix = new BucketAndPrefix(fileLocation);
        var bucketName = bucketAndPrefix.getBucketName();
        var prefix = bucketAndPrefix.getPrefix();
        var amazonS3Client = bucketAccountMap.get(bucketName);

        ListObjectsV2Result result = amazonS3Client.listObjectsV2(bucketName, prefix);
        List<S3ObjectSummary> objects = result.getObjectSummaries();
        return new ObjectSummaryResult(objects, amazonS3Client, bucketName);
    }

    @Getter
    private static class BucketAndPrefix {
        private String bucketName;
        private String prefix;

        BucketAndPrefix(final String fileLocation) {
            var split = StringUtils.split(fileLocation, "/");
            bucketName = split[0];
            prefix = split.length > 1 ? split[1] : null;
        }
    }

    @Getter
    private class ObjectSummaryResult {
        private AmazonS3 amazonS3Client;
        private String bucketName;
        List<S3ObjectSummary> objects;

        ObjectSummaryResult(List<S3ObjectSummary> objects, AmazonS3 amazonS3Client, String bucketName) {
            this.objects = objects;
            this.amazonS3Client = amazonS3Client;
            this.bucketName = bucketName;
        }
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/services/S3FileService.kt
package uk.gov.justice.digital.hmpps.riskprofiler.services

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.S3ObjectInputStream
import com.amazonaws.services.s3.model.S3ObjectSummary
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.FileType
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Viper
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.time.ZoneId

@Component
class S3FileService(
  @Qualifier("s3Client") private val s3Client: AmazonS3?,
) : FileService {

  override fun getLatestFile(fileLocation: String, fileType: FileType?): PendingFile? {
    val s3Result = getObjectSummaries(fileLocation)
    log.info("Found {} objects in {}", s3Result.objects.size, fileLocation)
    return s3Result.objects.stream()
      .max(Comparator.comparing { t -> t!!.lastModified })
      .map { o ->
        try {
          val s3Object = s3Result.amazonS3Client!!.getObject(s3Result.bucketName, o!!.key)
          return@map PendingFile(
            o.key,
            o.lastModified.toInstant()
              .atZone(ZoneId.systemDefault())
              .toLocalDateTime(),
            if (fileType == FileType.VIPER) {
              getViperFile(s3Object.objectContent)
            } else {
              s3Object.objectContent.delegateStream
            },
          )
        } catch (e: IOException) {
          return@map null
        }
      }.orElse(null)
  }

  override fun deleteHistoricalFiles(fileLocation: String) {
    val s3ObjectResult = getObjectSummaries(fileLocation)
    log.info(
      "Found {} data files for data housekeeping in {}",
      s3ObjectResult.objects.size,
      fileLocation,
    )
    s3ObjectResult.objects.stream().sorted(
      Comparator.comparing { obj: S3ObjectSummary -> obj.lastModified }
        .reversed(),
    ).skip(2).forEach { o ->
      s3ObjectResult.amazonS3Client!!.deleteObject(s3ObjectResult.bucketName, o!!.key)
      log.info("Deleted s3 data file: {} from bucket {}", o.key, s3ObjectResult.bucketName)
    }
  }

  private fun getViperFile(s3ObjectContent: S3ObjectInputStream): InputStream {
    val reader = BufferedReader(InputStreamReader(s3ObjectContent))
    var row: List<String>
    val rowList = mutableListOf<String>()

    while (true) {
      val line = reader.readLine() ?: break
      row = line.split(",")
      if (row[Viper.RECORD_ID] != "0") {
        rowList.add(line)
      }
    }
    s3ObjectContent.close()

    val csv = rowList.joinToString(System.lineSeparator())
    return csv.toByteArray().inputStream()
  }

  private fun getObjectSummaries(fileLocation: String): ObjectSummaryResult {
    val bucketAndPrefix = BucketAndPrefix(fileLocation)
    val bucketName = bucketAndPrefix.bucketName!!
    val prefix = bucketAndPrefix.prefix
    val result = s3Client!!.listObjectsV2(bucketName, prefix)
    val objects = result.objectSummaries
    return ObjectSummaryResult(objects, s3Client, bucketName)
  }

  private data class BucketAndPrefix(
    var bucketName: String?,
    var prefix: String?,
  ) {

    constructor(fileLocation: String) : this(null, null) {

      val split = StringUtils.split(fileLocation, "/")
      bucketName = split[0]
      prefix = if (split.size > 1) split[1] else null
    }
  }

  private data class ObjectSummaryResult(
    var objects: List<S3ObjectSummary?>,
    val amazonS3Client: AmazonS3?,
    val bucketName: String,
  )

  companion object {
    private val log = LoggerFactory.getLogger(S3FileService::class.java)
  }
}
